### PR TITLE
Lock cleanup refactoring

### PR DESCRIPTION
**Pull request by danielmarbach**

### CODE REVIEW

File: src/NServiceBus.Persistence.DynamoDB/SynchronizedStorage/ILockCleanup.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    using System.Threading.Tasks;
+    using Amazon.DynamoDBv2;
+
+    interface ILockCleanup
+    {
+        Guid Id { get; }
+
+        bool NoLongerNecessaryWhenSessionCommitted { get; set; }
+
+        Task Cleanup(IAmazonDynamoDB client, CancellationToken cancellationToken = default);
```
Comment (timbussmann) [inline src/NServiceBus.Persistence.DynamoDB/SynchronizedStorage/ILockCleanup.cs:14]: random thought: while against the guidance, I wonder whether we really need a cancellation token given it seems to be a type of action that should not be cancelable

**Review by timbussmann — APPROVED** [review]
just some naming thoughts, but happy to merge

File: src/NServiceBus.Persistence.DynamoDB/Saga/SagaPersister.cs [hunk 2]
```diff
@@ -217,8 +171,7 @@ public Task Save(IContainSagaData sagaData, SagaCorrelationProperty correlationP
             if (configuration.UsePessimisticLocking)
             {
                 // we can't remove the action directly because the transaction was not completed yet
-                var dynamoSession = (DynamoDBSynchronizedStorageSession)session;
-                dynamoSession.MarkSagaLockAsReleasedOnCommit(sagaData.Id);
+                dynamoSession.MarkAsNoLongerNecessaryWhenSessionCommitted(sagaData.Id);
```
[the remark this answers is not in the document]
Comment (danielmarbach) [reply src/NServiceBus.Persistence.DynamoDB/Saga/SagaPersister.cs:174]: Decided to add a param name for better readability

File: src/NServiceBus.Persistence.DynamoDB/SynchronizedStorage/DynamoDBSynchronizedStorageSession.cs [hunk 3]
```diff
@@ -75,6 +70,13 @@ public ContextBag CurrentContextBag
 
         public void MarkSagaLockAsReleasedOnCommit(Guid sagaId) => storageSession.SagaLocksReleased.Add(sagaId);
 
-        public void CleanupSagaLock(Guid sagaId, Func<IAmazonDynamoDB, CancellationToken, Task> cleanupAction) => storageSession.CleanupActions[sagaId] = cleanupAction;
+        public void Add(ILockCleanup lockCleanup) => storageSession.Add(lockCleanup);
```
[the remark this answers is not in the document]
Comment (danielmarbach) [reply src/NServiceBus.Persistence.DynamoDB/SynchronizedStorage/DynamoDBSynchronizedStorageSession.cs:73]: We decided to rename it to ` AddToBeExecutedWhenSessionDisposes`

File: src/NServiceBus.Persistence.DynamoDB/SynchronizedStorage/ILockCleanup.cs (hunk 1 above)
Comment (danielmarbach) [reply src/NServiceBus.Persistence.DynamoDB/SynchronizedStorage/ILockCleanup.cs:14]: We discussed this and concluded from an implementation perspective the cleanup actions can handle cancellation and it is a top level concern to decide whether an action is cancellable. Floating the token through seems to make sense.

**Pull request merged**